keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,727 @@
+
+Since swap provisioning is out of scope of this proposal, this enhancement poses little risk to Kubernetes clusters that will not enable swap.
+
+## Design Details
+
+### TL;DR
+
+In a nutshell, the following implementation are planned for Memory Swap Support
+in 1.22 GKE alpha
+
+1. Having a feature gate `SupportNodeMemorySwap` guarding against the memory
+   swap support feature
+2. Keep the default value of kubelet flag `--fail-on-swap` to `true` in order
+   to minimize the blast radius
+3. Introducing two new kubelet config `MemorySwapLimit` and `Swappiness`
+4. Introducing two new CRI parameter `memory_swap_limit_in_bytes` and `memory_swappiness`
+5. End to end wiring from kubelet config file to CRI
+
+### Expected User Behaviour
+
+For alpha, the feature gate `SupportNodeMemorySwap` is default to disabled, and
+`--fail-on-swap` flag value is the same as 1.21. Therefore, from Kubernetes
+user’s perspective, no behavior changes out of the box.
+
+For users that are ready to explore the Memory Swap feature in 1.22 Alpha, they
+will need to complete the following steps
+
+1. provision swap enable `SupportNodeMemorySwap` flag AND
+2. set `--fail-on-swap` flag to `false`
+
+Then, the user can start experimenting/fine tuning kubelet configuration
+`MemorySwapLimit` and/or `Swappiness` and observe the changes.
+
+### New Kubelet Configuration
+
+We will be introducing two new parameters to `KubeletConfiguration struct`
+defined in
+[https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go).
+These two configurations, if set, will apply to every container of the Node
+where kubelet is running.
+
+|Name|Description|Default Value|Feature Gate|
+|--- |--- |--- |--- |
+|MemorySwapLimit|This parameter sets total memory limit (memory + swap). This limits the total amount of memory this container is allowed to swap to disk.|-2, which enable disable swap|SupportNodeMemorySwap|
+|MemorySwappiness|This configuration sets how aggressively the kernel will swap memory pages. By default, the host kernel can swap out a percentage of anonymous pages used by a container. Users can set value between 0 and 100, to tune this percentage.|Unset, which will use host value|SupportNodeMemorySwap|
+
+#### MemorySwapLimit details
+
+MemorySwapLimit configuration is a kubelet flag that only takes effect on a
+container that has a memory limit set, either explicitly from
+[PodSpec]([https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits)
+) or implicitly from [Resource
+Quota]([https://kubernetes.io/docs/concepts/policy/resource-quotas/](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+).
+
+For container with memory limit set, MemorySwapLimit setting will have the
```

Comment:
Why limit swap usage to containers with memory limits? If `MemorySwapLimit` is set to -1, as an end user I'd expect my best effort containers would be able to use swap.

Reply:
This limitation essentially mostly comes from docker runtime, where `--memory-swap is a modifier flag that only has meaning if --memory is also set.` per [docker documentation](https://docs.docker.com/config/containers/resource_constraints/#--memory-swap-details). 

I think it makes sense in that we should protect the critical pods in production projects by setting limits so they are classified as Guaranted. BestEffort or Burstable pods should be used in developing projects only, per [CNCF blog recommendation](https://www.cncf.io/blog/2020/06/10/kubernetes-resources-management-qos-quota-and-limitrangeb/)